crates/core/c8y_api/src/smartrest/smartrest_serializer.rs

```diff
@@ -32,185 +93,124 @@ impl From<CumulocitySupportedOperations> for &'static str {
-        }
-    }
-
-    pub fn add_operation(&mut self, operation: &'a str) {
-        self.supported_operations.push(operation);
-    }
+pub fn declare_supported_operations(ops: &[&str]) -> String {
+    format!("114,{}", fields_to_csv_string(ops))
 }
 
-impl<'a> SmartRestSerializer<'a> for SmartRestSetSupportedOperations<'a> {}
-
 #[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
 pub struct SmartRestSoftwareModuleItem {
     pub software: String,
     pub version: Option<String>,
     pub url: Option<String>,
 }
 
-#[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
-pub struct SmartRestGetPendingOperations {
-    pub id: &'static str,
+/// A supported operation of the thin-edge device, used in status updates via SmartREST
+///
+/// This has two implementations, `&str` for custom operations, and [CumolocitySupportedOperations]
+/// for statically supported operations.
+pub trait C8yOperation {
+    fn name(&self) -> &str;
 }
 
-impl Default for SmartRestGetPendingOperations {
-    fn default() -> Self {
-        Self { id: "500" }
+impl C8yOperation for CumulocitySupportedOperations {
+    fn name(&self) -> &str {
+        (*self).into()
     }
 }
 
-impl<'a> SmartRestSerializer<'a> for SmartRestGetPendingOperations {}
-
-#[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
-pub struct SmartRestSetOperationToExecuting {
-    pub message_id: &'static str,
-    pub operation: &'static str,
-}
-
-impl SmartRestSetOperationToExecuting {
-    pub fn new(operation: CumulocitySupportedOperations) -> Self {
-        Self {
-            message_id: "501",
-            operation: operation.into(),
-        }
+impl<'a> C8yOperation for &'a str {
+    fn name(&self) -> &str {
+        self
     }
 }
 
-impl<'a> SmartRestSerializer<'a> for SmartRestSetOperationToExecuting {}
-
-#[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
-pub struct SmartRestSetOperationToSuccessful {
-    pub message_id: &'static str,
-    pub operation: &'static str,
-    pub operation_parameter: Option<String>,
+#[derive(Debug, Serialize, Eq, PartialEq)]
+#[serde(untagged)]
+pub enum TextOrCsv {
+    Text(String),
+    Csv(EmbeddedCsv),
 }
 
-impl SmartRestSetOperationToSuccessful {
-    pub fn new(operation: CumulocitySupportedOperations) -> Self {
-        Self {
-            message_id: "503",
-            operation: operation.into(),
-            operation_parameter: None,
-        }
+impl From<EmbeddedCsv> for TextOrCsv {
+    fn from(value: EmbeddedCsv) -> Self {
+        Self::Csv(value)
     }
+}
 
-    pub fn with_response_parameter(self, response_parameter: &str) -> Self {
-        Self {
-            operation_parameter: Some(response_parameter.into()),
-            ..self
-        }
+impl From<String> for TextOrCsv {
+    fn from(value: String) -> Self {
+        Self::Text(value)
     }
 }
 
-impl<'a> SmartRestSerializer<'a> for SmartRestSetOperationToSuccessful {}
+#[derive(Debug, Eq, PartialEq)]
+pub struct EmbeddedCsv(String);
```

Comment:
Additionally I've added this `EmbeddedCsv` struct for handling embedded CSVs (which come from custom operations with an result type of CSV. This will deserialize and reserialize the CSV when we add the fields to ensure a consistent format, and to ensure we don't generate a broken message (or multiple messages if the output contains newlines).